File: 109-有序链表转换二叉搜索树.py
class Solution:
    def sortedListToBST(self, head: ListNode) -> TreeNode:
        def sortedlist_to_bst(head, tail):
            if head == tail:
                return None
            
            fast = slow = head
            while fast != tail and fast.next != tail:
                fast = fast.next.next
                slow = slow.next
            
            root = TreeNode(slow.val)
            root.left = sortedlist_to_bst(head, slow)
            root.right = sortedlist_to_bst(slow.next, tail)
            
            return root

        return sorted_list_to_bst(head, None)

File: 109-有序链表转换二叉搜索树.cpp
class Solution {
public:
    TreeNode* sortedListToBST(ListNode* head) {
        vector<int> myvector;
        
        while(head){
            myvector.push_back(head->val);
            head = head->next;
        }

        return sortedarray_to_bst(0, myvector.size(), myvector);
    }

    TreeNode* sortedarray_to_bst(int left, int right, vector<int>& myvector){
        if(left == right)
            return NULL;
        
        int middle = (left + right) / 2;
        TreeNode* root = new TreeNode(myvector[middle]);

        root->left = sortedarray_to_bst(left, middle, myvector);
        root->right = sortedarray_to_bst(middle + 1, right, myvector);

        return root;
    }
};
